### ISSUE TITLE
Adjustments to crits and hard mode stat exp

### ISSUE THREAD

**Issue by DiabloStorm** [jojobear13/shinpokered#234]
<details>
  <summary>Attempt 1</summary>
  
### EV progression

Early game team (and their evolutions):
Charmander **(med-slow exp growth)** (L:5)
Pikachu **(med-fast exp growth)** (Caught at L:3?)
Spearow **(med-fast exp growth)**(Caught at L:5?)
Abra **(med-slow exp growth)**(Caught at L:12)
Meowth **(med-fast exp growth)**(Caught at L:10)
Diglett **(med-fast exp growth)**

**Using non-randomized wild pokemon, randomized trainer pokemon, trainer scaling, set mode (non !), and intending to battle every trainer at least once**

****Up to Viridian Forest**** : (Some baselines before I progress even more)
Spearow(L;12): HP:1334, A:994 D:1290, SPD:1455, SPC:855
Charmander(L:11): HP:738, A:992, D:765, SPD:1279, SPC:622
Pikachu(L:11): HP:1724, A:2258, D:1755, SPD:2887, SPC:1377

****After Pewter Gym and the remaining 2 Bug Catchers****:
Spearow(L;13): HP:1432, A:1088 D:1410, SPD:1552, SPC:970
Charmander(L:15): HP:1001, A:1314, D:1145, SPD:1458, SPC:847
Pikachu(L:12): HP:1790, A:2330, D:1848, SPD:2962, SPC:1457

****Route 3****: L:15 Update
Spearow(L;15): HP:1648, A:1327 D:1658, SPD:1767, SPC:1219
Charmander(L:15): HP:1031, A:1341, D:1170, SPD:1480, SPC:867
Pikachu(L:15): HP:2031, A:2663, D:2166, SPD:3244, SPC:1705

****After Cerulean Rival Battle****
Spearow(L;21): HP:2362, A:2168 D:2451, SPD:2704, SPC:2047
Charmander(L:20): HP:1419, A:1795, D:1563, SPD:2011, SPC:1293
Pikachu(L:21): HP:2757, A:3609, D:3098, SPD:4213, SPC:2569

</details>

### Hard Mode EV Progression
  
  <details>
  <summary>Playthrough Variables</summary>
  
Non-randomized wild pokemon, randomized trainer pokemon, trainer scaling, set mode on, Battle Style ***!*** on, SLP/FRZ clauses off, Item Clause on, Low level pokemon catch-up exp **_OFF_** (also, not that I want to but...) skipping trainers is a must due to the detriment they have on EV gains...unless I need the levels or have to fight them (min-maxing to survive does not always make for the most fun playthrough)

</details>
  
Charmander **(med-slow exp growth)** (L:5) (Rerolled until near perfect DVs, I'm getting those 5 pokeballs this time)
Pikachu **(med-fast exp growth)** (Caught at L:3)
Spearow **(med-fast exp growth)**(Caught at L:3)
 ~~Abra **(med-slow exp growth)**(Caught at L:tbd)~~
 ~~Meowth **(med-fast exp growth)**(Caught at L:tbd)~~
 ~~Diglett **(med-fast exp growth)**(Caught at L:tbd)~~

<details>
  <summary>Rival Battle 1 Note</summary>

Charmander DVs: 14, 15, 13, 15, 14. This game mode awards too much exp so I deliberately lose the first battle to avoid gaining 2 levels which would actually put me at a disadvantage for the next rival battle. (Seems counterintuitive to play this way, but I'd gain more EVs during these 2 levels by fighting level 2 Nidoran♂. The "1.5x" EVs from the trainer battle amounts to nothing)

Maybe the bonus EXP from trainer battles is just a bad idea combined with the other changes.

</details>

<details>
  <summary>Strategy for Oak's 5 free Pokeballs?</summary>
  
  (I'm pretty sure this wont work AT ALL with Squirtle or Bulbasaur as starter!!)
  
I've tried to EV train Charmander on level 2 Nidoran♂ until level 11 to hopefully avoid squirtle getting water gun but was overpowered by pidgey.

Assuming The Rival here has +0-3 levels more than you it looks like I'd have to be somewhere between level 18 and 19 just through training on around 270 level 2 Nidoran♂ to have about the same EVs as Rival (in attack anyway, because of the 640 spread he likely has more of everything else)...which I doubt would be enough anyway since he has 2 pokemon.

Doing some quick math, Leveling to 30 grants Slash and maxes out my attack EVs by fighting level 2 Nidoran♂, so that's the next attempt.
...After finding and fainting 1280 level 2 Nidoran♂...
...This is ridiculous just to obtain 5 free pokeballs, not going to lie. ~~It also might not even be possible with item clause on.~~
Considering the time investment this takes, I'm just going to artificially train Charmander this way in PKHex to save time and see if there is any success in battle.
Setting Charmander's exp to 21895 (Level 5 has 135xp already technically) 1280 level 2 Nidoran♂ battled, 17 exp each)
EVs from these Nidoran♂:
HP:46, A:57, D:40, SPC:40, SPD:50 
x1280
= HP:58880, A:65535, D:51200, SPC:51200, SPD:64000

[So, after what amounts to probably 2+ days of playtime, hunting out and fainting 1280 level 2 Nidoran♂ and using the hopelessly broken crit mechanics with Slash, you too can obtain Oak's 5 free pokeballs. Oh, also you're not allowed to evolve.](https://www.youtube.com/watch?v=PT3I8EHqhf4)

  </details>

### After Rt. 22 Rival battle 2: (TL;DR: Rival has an impossible amount of EVs if you're lower than level 19 in Viridian and you will not win with just an unevolved starter until level 30)

  "Hard Mode"is scuffed. I admit I was wrong, it isn't impossible to get the 5 Pokeballs from Oak. It's just that the process is absurd and not fun and I highly doubt this is what entails a "true" experience. You haven't fixed the problem of overleveling and powering through the game, you've merely shifted the goalposts from levels to EVs.
  I'm continuing the playthrough at level 31 since that is what it takes to do the mini quest for the free pokeballs. Fainting 1280 level 2 Nidoran♂ isn't hard. This is called tedium.
  
 Charmander(L:31): HP:59002, A:65535, D:51335, SPD:64129, SPC:51315
 
### After Pewter Gym:

[Battle here](https://www.youtube.com/watch?v=pMoPusdWh0M)

Charmander(L:33): HP:65535, A:65535, D:65535, SPD:65535, SPC:65535 (Trained on 1635 total level 2 Nidoran♂ and one battle w/ the last viridian bug trainer)
Fearow(L;30): HP:65535, A:65535 D:65535, SPD:65535, SPC:65535 (Trained on 1639 level 2 Nidoran♂)
Pikachu(L:31): HP:65535, A:65535, D:65535, SPD:65535, SPC:65535 (Trained on 1639 level 2 Nidoran♂)

...You get the idea...

Disclaimer: I caught pokemon in game. checking DVs in game but I don't have the patience to torture myself training on this many cherry picked Nidoran♂. I reproduced a legitimate EV training result via PKHeX, anybody can do this purely in-game but you'll need a psych ward afterwards because you'll be seeing the same things on screen literally tens of thousands of times. Or maybe that sounds like fun to some people? Not me.

### Summary

Comparing to my first attempt, I see no point in continuing to log EVs. The game is flawed and formulaic.  I don't normally play pokemon this way and I didn't have to learn and adopt this playstyle until this game forced me to.
The goal posts have simply been moved from overleveling to power through, to excessive EV training early to power through while keeping your team level as low as possible. It's not hard, it's simply repetitive and time consuming and not necessarily better or "truer" due to the grind in the beginning.

The way to succeed in "Hard Mode" requires an absurd setup in that you max out EVs first thing whenever you get new pokemon. (The lower level the caught pokemon intended for use, the better, as you have more time to EV train without leveling too high.)
Max out EVs on level 2 Nidoran♂ in Rt 22, pidgey north of Pallet town, whatever. It's basically mandatory to survive. 

The process isn't fun, would probably take several days in real time per pokemon, but once that's out of the way, you can maybe go ahead and fight some trainers for fun again and use them to gain EXP as normal without much detriment. (Factoring in that they catch up to your EVs by level 100, you're still generally discouraged from battling tons of trainers if you wish to breeze through the game at a minimal level where the differences between yours and the AIs EVs are greatest.)

You simply repeat the process until you beat the game. Since the AI is locked to a linear amount of 640 EV gains per level, by around level 30 (varies depending on pokemon caught-at level) your pokemon have the same EVs as them at level 100, why their EVs aren't just tied roughly to your own EV progression in some way, I have no idea. Too complicated? 
This would remove the need for such a grind and maintain some level of fairness and "challenge".
 I say "challenge" because....

### Crit rate
***Roulettemon.*** Another thing squeezing your playstyle into a little box. You can't beat the crit system so might as well join it. (Hence my pokemon selection, though there are many other options including new ones that didn't originally exist thanks to the dire hit fix which makes the crit system even more pervasive than originally, enabling slower 'mons to always crit with high crit moves) 
<details>
  <summary>Summary List of Critical Attackers (With some newly existing with >100% crit due to this romhack)</summary>

Farfetched with Slash (Base speed 60) = 93.75%. With Direhit = >100% **(exclusive to this hack)**

Kabuto with Slash (Base speed 55) = 85.94%. With Direhit = >100% **(exclusive to this hack)**

Kabutops with Slash = >100%

Pinsir with Slash = >100% 

Paras (Base speed 25) with Slash = 39.06%. With Direhit = >100% **(exclusive to this hack)** (Spore for up to 7 free turns of setup!)

Parasect (Base speed 30) with Slash = 46.88%. With Direhit = >100% **(exclusive to this hack)** (Spore for up to 7 free turns of setup!)

Sandshrew with Slash (Base speed 40) = 62.5%  With Direhit = >100% **(exclusive to this hack)**

Sandslash with Slash  = >100%

Bulbasaur with Razorleaf (Base speed 45) = 70.31%. With Direhit = >100% **(exclusive to this hack)**

Ivysaur with Razorleaf (Base speed 60) = 93.75%. With Direhit = >100% **(exclusive to this hack)**

Venusaur with Razorleaf = >100%

Bellsprout with Razorleaf (Base speed 40) = 62.5%. With Direhit = >100% **(exclusive to this hack)**

Weepinbell with Razorleaf (Base speed 55) = 85.94%. With Direhit = >100% **(exclusive to this hack)**

Victreebel with Razorleaf = >100%

Mankey with Karate Chop = >100%

Primeape with Karate Chop = >100%

Machop with Karate Chop = 54.69%. With FocusEnergy/Direhit = >100% **(Direhit saves a moveslot exclusively in this romhack)**

Machoke with Karate Chop = 70.31%. With FocusEnergy/Direhit = >100% **(Direhit saves a moveslot exclusively in this romhack)**

Machamp with Karate Chop = 85.94%. With FocusEnergy/Direhit = >100% **(Direhit saves a moveslot exclusively in this romhack)**

Krabby with Crabhammer (Base speed 50) = 78.13%. With Direhit = >100% **(exclusive to this hack)**

Kingler with Crabhammer = >100%

Charmander with Slash = >100%

Charmeleon with Slash = >100%

Charizard with Slash = >100%

Meowth with Slash = >100%

Persian with Slash = >100%

Scyther with Slash = >100%

  </details>

ANY pokemon with base speed of 25 or more (ALL but 3) that have used a DireHit essentially now have base 100 speed or more  in consideration of crit chance calculation. This puts their normal attacking moves all at 19.53% or more chance to crit. Similar to your "crit-zilla" Tauros. 
Any pokemon >=130 Base Speed or more (4 pokemon) with Direhit have >100% crit chance with ANY move.
Any pokemon >=65 Base Speed (ALL but 67) with Direhit have >50% crit chance with ANY move.
I don't know if the AI uses Direhit, but this is yet more evidence of increased prevalence exclusive to this romhack.
  
  
Overall the utilized crit rate system reduces the game to one of chance rather than any challenge. So I went all in with Double Team (roulette fighting roulette, the only way to avoid getting crit), "high crit chance" moves (essentially 100%) and high speed 'mons.
The way crit works makes boosting moves irrelevant, it cuts right through defense/special boosts used by the opponent and ignores any boosts you have so why would you bother using those boosts? Throw them right in the trash. Yes, I understand that's how all pokemon games have been but the issue here is the sheer amount of critical hits you'll see during a playthrough. I frequently use boosting moves in the normal games where everything works under a more rational 1/16 system for crits.

The problem only gets worse as you increase in levels since crit damage is calculated as your pokemon being double the level it is. Trainer scaling and fixes to direhit/focus energy and to an extent, random trainer pokemon make the problem ever more prevalent throughout gameplay, it becomes way more apparent than in the original game. After you're forced to EV train to survive and are around level 30, most of the trainers are using evolved pokemon.

I would normally welcome the fixes to direhit/focus energy and additions like trainer scaling and random trainer pokemon, I love the ideas. They are no brainers. The problem is they only add more frustration towards the old crit system being used because they further increase the prevalence of crits experienced over the original game. Changes like this would shine more if there wasn't refusal to switch to a 1/16 crit system.

### TL;DR:

- You've made winning trainer battles punishing rather than rewarding. (Has nothing to do with challenge, read full version.)
- Gaining bonus EXP (including from in-game trades with the cpu or through Link Cable Club) handicaps those pokemon.
- End game will be easier to win if you deliberately lose in battles that the game doesn't demand you win in order to progress and avoiding battles entirely when possible so that you can stay as low level as possible with max EVs.
- Please fix the stupid crit system to 1/16 and high crit moves to 1/8 to remove excessive roulette factor. There's a reason why they dumpstered the gen 1 crit system and never went back to it.
- Lock AI EVs to a percentage (RNG: 80-110%?) of the player team's EVs to eliminate grueling grindfests that are mandatory straight from the start of the game. (Don't apply this solution to Rival battle 2, this needs special consideration because of Oak's mini quest.)
- >Shiny Mercy
When a player encounters an AI trainer shiny pokemon, the next wild encounter will be shiny
AI trainers can only have shinies in hard mode, so it affords more chances to find wild shinies

Shinies are worthless in hard mode due to their inferior DVs, yet, nonsensically, you've made a restriction to enable this only in hard mode.

**Comment by jojobear13** [comment]
Version used? Is it still a problem in [v1.23.11](https://github.com/jojobear13/shinpokered/tree/master_dev/patches_and_info)?

**Comment by jojobear13** [comment]
> Additionally, My pokemon (different OT/ID)..all under level 15 (but above 10), are disobeying even though this option is turned off.

The obedience cap option is specifically for applying obedience to all pokemon using an adjusted scale. If this option is off, then the game reverts to how obedience works in the vanilla games. In the vanilla games, obedience is only applied to pokemon with a different OTID and has caps as follows:
`ObedienceLevelsTraded:
	db 10	;no badges
	db 0	;boulder badge
	db 30	;cascade badge
	db 0	;thunder badge
	db 50	;rainbow badge
	db 0	;soul badge
	db 70	;marsh badge
	db 0	;volcano badge
	db 255	;earth badge`

**Comment by DiabloStorm** [comment]
> > Additionally, My pokemon (different OT/ID)..all under level 15 (but above 10), are disobeying even though this option is turned off.
> 
> The obedience cap option is specifically for applying obedience to all pokemon using an adjusted scale. If this option is off, then the game reverts to how obedience works in the vanilla games. In the vanilla games, obedience is only applied to pokemon with a different OTID and has caps as follows:
> 
> * db 10	;no badges
> * db 30	;cascade badge
> * db 50	;rainbow badge
> * db 70	;marsh badge
> * db 255	;earth badge

Thanks for clarification. Also, I tried but wasn't able to reproduce the incident with growl "failing" though I do remember looking at DVs and stat experience prior...

I'm on 1.23hotfix ...im not sure if that is 1.23.11.... though, where are the bps patches so I can confirm? I admittedly downloaded this from somewhere else because I couldn't find simple patches and compiling was having me download gigabytes of programs i'll never use again. Is there a simple file patch? It's mentioned in the description but I don't see it.


**Extra crap to read if you have the time:**
----------------------
Additional info: ..unsure if this is a creepy easter egg (you should totally do that for something hidden pretty well) or if you were trying to remove trainer escape glitch in this hack entirely (I notice you specifically mentioned "Can't use surf/teleport/escape rope to escape from trainer encounters"is it deliberate that you left methods? there are other ways)

I found this
![pay respects](https://user-images.githubusercontent.com/63818362/209598755-e3209fef-56ad-4063-977b-f1d87677d54a.png)

Also also...(a question/suggestion for something I haven't gotten far enough to see in game but read about in the description) why is strength quick key default to no interactable object instead of fly? strength quick key would make more sense if you're facing a boulder *only* and fly if nothing else. Just my opinion there.

And, I can understand wanting to change moves like wrap (again something I haven't experienced in game yet but read here about so maybe I don't understand something) but... 
"Each time a trapping move is used and hits, user's speed reduced 25% until recalculated through other mechanics"
isn't it enough to just cost a turn loss when you switch out? Why a stacking speed debuff when the move does damage? Is it each time damage is dealt? Because that makes trapping moves entirely unappealing and by the description which I possibly don't understand...a bit extreme.

But anyway, overall immensely impressive work. I have to debate whether to show this or the originals to someone new to pokemon

**Comment by jojobear13** [comment]
> I'm on 1.23hotfix ...im not sure if that is 1.23.11.... though, where are the bps patches so I can confirm? I admittedly downloaded this from somewhere else because I couldn't find simple patches and compiling was having me download gigabytes of programs i'll never use again. Is there a simple file patch? It's mentioned in the description but I don't see it.

1.23.11 is the current beta for the upcoming 1.24 patch (due sometime next year). A lot of things have been fixed already. Patch files for this beta version are found in the master_dev "patches and info" folder here: https://github.com/jojobear13/shinpokered/tree/master_dev/patches_and_info

Click on a patch file. It will take you to another screen. Click the download button to the right of the screen. Use a program like FLIPS to apply the patch to the correct ROM file. BPS patches use a hash verification so it won't let you mess up.

>Now, I know you said you didn't want to decouple it from speed, but there's a randomizer program that ive used (PokeHaxRandomizer) that edits roms to change it to 1/16 like other games, 

**And there's your problem right there**. External randomizers SHOULD NOT be used on this rom hack. Why? because too much has changed. The randomizer is editing values in rom addresses that do not correspond to the vanilla release. This will cause all sorts of unpredictable behavior. And no, seeing how a an external program interacts with a rom is not an easy thing to do.

The master_dev 1.23.11 beta that I linked previously has a randomizer built into the game itself.

> Edit: on the off chance... if pidgey were to use sand attack and growl misses...does the game print it as "fails" still in this hack? I don't remember it using sand attack at all though which is why this baffled me.

The text will be "[pokemon] used [move]! But, it failed!" This is the same as in the vanilla games. 

>why is strength quick key default to no interactable object instead of fly? strength quick key would make more sense if you're facing a boulder only and fly if nothing else. Just my opinion there.

It's not that simple. Fly needs to go through a menu because of how the town map gets loaded. Calling it from the field causes corruption. I would love to make Strength conditional on facing a boulder, but boulders exist on the object layer instead of the background layer. Objects have a lot more settings a fiddly bits that makes identifying them in code a pain.

>"Each time a trapping move is used and hits, user's speed reduced 25% until recalculated through other mechanics"
isn't it enough to just cost a turn loss when opponent switches out of these moves? Why an additional stacking speed debuff to the user when the move does damage? Is it each time damage is dealt?

It's each time the move's effect gets initialized and the number of attacks are determined; misses don't count and hits within the trapping effect don't count. This debuff was added because the improved AI on certain trainers were creating unwinnable matches in the master branch. It's possible that Ekans and Bellsprout trainers could outspeed your entire team, making the free switch useless, and wrap spam you endlessly. This happened a few times in playtesting and it was awful. This debuff puts a degradation on wrap spam that will eventually allow the player to get a hit in edgewise after two to three rounds of wrap spam.

**Comment by DiabloStorm** [comment]
> **And there's your problem right there**. External randomizers SHOULD NOT be used on this rom hack. Why? because too much has changed. The randomizer is editing values in rom addresses that do not correspond to the vanilla release. This will cause all sorts of unpredictable behavior. And no, seeing how a an external program interacts with a rom is not an easy thing to do.

The randomizer doesn't even open your rom hack, so you're mistaken there.
![image](https://user-images.githubusercontent.com/63818362/209615641-9730799d-65da-459d-8781-7977609a9190.png)

**Comment by jojobear13** [comment]
If that's the case, then are you playing in SET mode? If so, then the game is in Hard Mode. AI trainer pokemon are going to have better stats and better speed due to DVs and stat EXP (Lorelei's Slowpoke CAN outspeed you on occasion due to sheer stat EXP). This translates to more crits. Version 1.23.11 separates hard mode into its own option for those who do not want trainers to have stat exp and better DVs. But still, vanilla gen 1 has half the 'dex turn into crit-zillas in the late game. It's why Tauros was such a beast.

That graphical glitch is caused by an invalid address jump somewhere. I think I fixed that already. I've tried to close as many holes and escape glitches as I can, but once in a while a new one gets discovered.

**Comment by DiabloStorm** [comment]
Thanks for all the responses, I am now on the proper version (not used to github's layout tbh, but I see where they are now) Which seems to have fixed an issue I hadn't reported but was going to.

Yes I am on set mode btw.

Unfortunately the state I had for that text box glitch image was with the old version, but I trainer escaped the last trainer in viridian and then talked to the old man, did the tutorial, some crazy effects ensued when returning to the forest (Can't confirm this remains yet but now I'm curious) I'll test again and submit in a new Issue if this is unintended?

Are trainer escapes supported with this hack or is the goal to remove them? (for future reference)

**Comment by jojobear13** [comment]
The goal is to remove trainer escapes. [Insert "No Fun Allowed" meme].

The calculation for gen 1 crits is as follows: X = RoundDown(0.5 * base speed). X = X * 4 if using dire hit or focus energy. X = X * 8 if using a high crit move. X is capped at 255. Critical hit achieved if a random number [0, 255] is < X.

So it only really scales with base speed, but the fact that crits are rolled on a 256-sided die means there is a high degree of variance in the results (sigma value of 73.612 under ideal circumstances). So Shellder has a crit rate of 7.8% *on average* but this can easily skew up or down by a large amount based on how the RNG feels about you that day. Compare this to Charizard at 19.5% (that's before Slash boosts it up to 100%). Or Jolteon at 25.4% (crits 100% on everything after a single Focus Energy). 
 
The RNG for these games is not particularly great. I have tried to replace the random number engine several times with something that is "good enough". Even still, the time between button presses is the main source of external entropy. Hammering on the A button, even when reloading states, can work things into a pattern. If your emulator generates random garbage for the RAM at startup (like BGB does to emulate a real gameboy), then every startup will give a new RNG seed.

**Comment by DiabloStorm** [comment]
Well, I was thinking about it last night, the viable option in the current state is to run an evasion team...which doesn't sound very fun at all. I really didn't want to change off of "SET" ...the difference in crits is negligible. _(By the way what is the new **!** in the battle style options? )_

Crits don't make the game any more challenging than saying "roulette is challenging". It's cheap and I'm finding that it works in the AI's favor more than my own. Additionally...My super effective moves aren't even really doing that much damage, yet (even though ai usually has +3 levels) theirs usually (or just about) take me out in a single hit, "not very effective" does about as much damage to me as my normal effective moves for them. (This is the general feel overall)

I'm assuming you had help with porting over the more deleter/relearner from the other hack? Otherwise if you are able to pull things from other hacks, how is that any different from pulling changes from a rom that was edited via randomizer?. Theoretically couldn't you compare a base rom to one that only has this activated to see what they did? (Yes, easier said than done I'm sure)
![image](https://user-images.githubusercontent.com/63818362/209689383-dd724264-5ce8-485a-9c7a-a77c83c90aab.png)
Or... maybe (Dabomstew, who made that randomizer) would just simply show you how.

I can see a post reply from you to another person saying a lot of the same stuff as me about crit back in 2020 to which **you** say "Critical hits being tied to speed means that fast crit-seekers can easily sweep a whole team. There is no countermeasure to this unless you get lucky and can set up evasion spam."
This certainly still seems to be the case. Was that the intention? Because this doesn't seem avoidable. Just my opinion but I'm not sure how this would be considered "fun" or even a good design for the "true" versions.
You even mention "**Critical hit rate for enemies is slightly higher than vanilla because their speed stat is boosted by stat exp.** The Gen 1 RNG is also not the best, and it seems to be seeded by the time between button inputs."

Context: https://www.pokecommunity.com/showthread.php?p=10110801#post10110801

Which admittedly by you, you have deviated from the original game's system *anyway* with enemy stat exp which has entirely thrown off the original crit rate already. Just by playing, it certainly isn't "slightly" off.

I've played through these games many times on original hardware, then with randomizers and now with your hack, while very impressive, the crit mechanics are always going to hang over this one like a fog. If you've played with the randomizer I've shown, you'd see that it's a lot more reasonable. That, paired with the ai changes you have in this rom would, to me, be the sweet spot.  It's unfortunate if that were to never be a consideration. 

I was just in the beginning of another randomizer run but noticed your hack early enough and saw that it did what I wanted the game to do, implemented in a much better way and switched over but the crits are really untenable. Again...unless I plan to play as an evasion team or I save state after every single move to eliminate crits completely which would be absolutely ridiculous ways to attempt to enjoy your game.

**Comment by jojobear13** [comment]
I frequently get confused how critical hits are calculated due to how many changes were made over each generation (or even game to game). Stat exp or any boosts will not effect crits. Crit rate is tied to base speed, and is determined by species. Anything else is just me smoking crack, implementing something totally wrong, or beta experimentation.

> Otherwise if you are able to pull things from other hacks, how is that any different from pulling changes from a rom that was edited via randomizer?

It's the difference between binary hacking (which is what randomizers do) and decomp hacking (which is what I do). When I pull things from other hacks, I'm re-implementing them entirely in source-code rather than modifying individual bytes at specific addresses. So adding features from a randomizer is essentially starting with the general concept and working from there. No need to reverse-engineer.

>By the way what is the new ! in the battle style options?

Hard mode. It activates stat EXP and other stuff that was previously locked to SET. Press Left to toggle the battle style without toggling Hard mode. Press Right to toggle the battle style while also toggling Hard mode. It is now possible to play in set mode again per the vanilla games. Turning off hard mode will make those trainer pokemon less chunky.

This is the function for determining critical hit chance in both:
- Vanilla: https://github.com/pret/pokered/blob/53805947c3bc30c491b4f2a820c27e152dfa4b89/engine/battle/core.asm#L4598
- and Shinpokered: https://github.com/jojobear13/shinpokered/blob/2523b0630d28f64f99beb6bdb4d2c893e0d4b634/engine/battle/core.asm#L5109

They are effectively the same. Moving to the Gen 2 crit chance mechanic is out of scope for this project. 

That said, there is something weird going on with the RNG in this function that is causing random numbers to skew lower than the expected mean. I will need to do some more experiments with the BattleRandom function and try a few things in the CriticalHitTest function. For example, I don't know why Gamefreak's Morimoto felt the need to bit-roll the random number three times, but this could be affecting something.

**Comment by DiabloStorm** [comment]
> Hard mode. It activates stat EXP and other stuff that was previously locked to SET. Press Left to toggle the battle style without toggling Hard mode. Press Right to toggle the battle style while also toggling Hard mode. It is now possible to play in set mode again per the vanilla games. Turning off hard mode will make those trainer pokemon less chunky.

Good idea separating all of those options. 
I'm okay with chunky, even the way it feels as if AI mon's are tanking my hits and dealing increased damage. I like adding challenge to the game but I just can't enjoy the roulette factor with crits as they are currently. I think most people find that they dislike things happening to them outside of their control so often.
With that said, thanks for bearing with me and reading through all of that, and for explanations.

I'm currently starting over with a new run...skewed towards speed and crit to see exactly how that...goes...being a bit compartmentalized and limited into the list of 23 pokemon I mentioned above, so...not ideal. 6 of which I can get early (I did notice the girl in Oak's lab added a presumably wild encounter randomizer which is a nice thing I didn't notice before)
We'll see how far Charmander, Pikachu, Abra, Meowth, Diglett and Spearow get me in early game, all will be equipped with double team. 
A new crit meta for now I suppose.

**Comment by jojobear13** [comment]
I have a hunch on something. Try re-patching with one of these patch files. See if if it helps any.
[crit debug patches.zip](https://github.com/jojobear13/shinpokered/files/10310872/crit.debug.patches.zip)

**Comment by DiabloStorm** [comment]
> I have a hunch on something. Try re-patching with one of these patch files. See if if it helps any. [crit debug patches.zip](https://github.com/jojobear13/shinpokered/files/10310872/crit.debug.patches.zip)

Will do. I'll continue this run from this patch

**Comment by DiabloStorm** [comment]
Question. Is PKHeX safe to use with this if we're trying to squash potential bugs? That program does work. (Not a randomizer if you aren't familiar. it's just a save file editor essentially, which I'm thinking is different, correct me if I'm wrong.)

**Comment by jojobear13** [comment]
Shinpokered does not mess with the save structure. Save editors are fine.

**Comment by DiabloStorm** [comment]
> Btw, this is with a 14 special DV Pikachu... What DVs are trainers rolling with?
Thundershock vs rival's pidgey:
image

Rival's 1 Gust:
image

Is this normal? lol All non crits.

I'm trying to understand the huge variance between these 3 fights I just did. (There are crits there too but that's not what I'm wondering about this time)
https://www.youtube.com/watch?v=zAVrEK30vAc

What are the rival's pokemon's DVs and EVs? I don't stand a chance 2v2 without a lot of luck or not until he's the same level, the first two tries he's slightly over my level but look at the gust bs thundershock damage, and the overall damage/bulk of the AI...is this right?

**Comment by jojobear13** [comment]
Hard mode is not balanced around having perfect stats. But it is balanced around having above-average stats and access to good move pools.

Hard mode does the following:
-   Enemy trainer pokemon are assigned 640 stat exp per level in each stat (they have zero otherwise)
-   Enemy trainer pokemon have randomized DVs that are 8 to 15 (otherwise they are always 9,8,8,8)
-   Badge-granted stat boosts are disabled in trainer battles
-   X-stat items have double the effect like in more recent generations
-   Revive items cannot be used in battles
-   Mewtwo will prevent you from using a master ball on it and use AI to choose moves
-   Hard mode gives an additional 1.5x EXP boost in trainer battles

Given a level 13 pidgey:
- Increasing a DV from 8 to 15 will give +1 point to a stat.
- A stat exp value of 8320 will result in 2 to 3 extra points in each stat. 

I'm open to adjusting the stat exp formula to make it less linear

**Comment by DiabloStorm** [comment]
> Hard mode is not balanced around having perfect stats. But it is balanced around having above-average stats and access to good move pools.
> 
> Hard mode does the following:
> 
> * Enemy trainer pokemon are assigned 640 stat exp per level in each stat (they have zero otherwise)
> * Enemy trainer pokemon have randomized DVs that are 8 to 15 (otherwise they are always 9,8,8,8)
> * Badge-granted stat boosts are disabled in trainer battles
> * X-stat items have double the effect like in more recent generations
> * Revive items cannot be used in battles
> * Mewtwo will prevent you from using a master ball on it and use AI to choose moves
> * Hard mode gives an additional 1.5x EXP boost in trainer battles
> 
> Given a level 13 pidgey:
> 
> * Increasing a DV from 8 to 15 will give +1 point to a stat.
> * A stat exp value of 8320 will result in 2 to 3 extra points in each stat.
> 
> I'm open to adjusting the stat exp adder to make it less linear

When you say "stat exp" Are you referring to EVs?

So, the enemy AI gains 640 to all EVs each pokemon level? This seems like an impossible amount to obtain this early...or at all. You may have simplified the programming by dividing 65535 into 100 levels but this is certainly not obtainable by the player at such a linear level. My pikachu is sitting at around 2k leveling up around viridian and the forest.

Seems like maybe a playtest nightmare. Do you need some sample EV numbers reported as I go through the stages of the game? Do trainers and rebattles award EVs (stat exp) btw?

If I turn off the **!** I don't get to battle Mewtwo with AI? That's a bummer.
X-stat items...double the effectiveness for all or just enemy trainers that use them?

Also, I take it virtually zero people are going to be able to fulfill the criteria needed in the early game to get the 5 free pokeballs from Prof Oak with the way this is all programmed. Absolutely nobody is going to beat Rival Battle 2 with just their starter on hard mode as it is.

**Comment by jojobear13** [comment]
>When you say "stat exp" Are you referring to EVs?

Yes, though EVs as they are properly called don't exist until Gen 3.

>Seems like maybe a playtest nightmare. Do you need some sample EV numbers reported as I go through the stages of the game?

Yes please.

>Do trainers and rebattles award EVs (stat exp) btw?

Yes, per the vanilla games. Trainer battles award 1.5x EVs.

>Also, I take it virtually zero people are going to be able to fulfill the criteria needed in the early game to get the 5 free pokeballs from Prof Oak with the way this is all programmed. Absolutely nobody is going to beat Rival Battle 2 with just their starter on hard mode as it is.

It's been done. Some people are...something.

>If I turn off the ! I don't get to battle Mewtwo with AI? That's a bummer.

Turn it back on in the unknown dungeon. Most of the difficulty options only affect trainer battles. Mewtwo is wild.

>X-stat items...double the effectiveness for all or just enemy trainers that use them?

For all.

>Additionally: "Hard mode gives an additional 1.5x EXP boost in trainer battles"
To most people they'd say "yay" ...not knowing that this is actually a detriment if you are on hard mode and using trainer level scaling.

Mixing those two is supposed to be brutal. It's more appropriate for challenging your optimized late-game team or a run through New Game +.

**Comment by DiabloStorm** [comment]
So, Maybe every 10 levels I should update an EV report? Maybe every 5? The problem with that is, depending on the pokemon, they could grow at different rates (with slower gaining more EVs per level). Or perhaps update per area? Maybe I'll do a mixture of the two. I'll include the pokemon "caught at" levels too (which is why Charmander is so much further behind in EVs than Pikachu)

**Let me know if you have any suggestions on how to best capture this data and reduce chaos.**

I'll find some way to edit this with my progression without it being too messy:


My theoretical early game team will consist of (and their evolutions):
Charmander**(med-slow exp growth)** (L:5)
Pikachu**(med-fast exp growth)** (Caught at L:3-5)
Spearow**(med-fast exp growth)**(Caught at L:3-5)
Abra**(med-slow exp growth)**
Meowth**(med-fast exp growth)**
Diglett**(med-fast exp growth)**

**Comment by DiabloStorm** [comment]
Update on crits:

Level 12 Exeggcute (base 40 speed) Barrage:
84 hits, 8 crits, 8 misses
Expected rate: 7.81% Observed: 8%

Level 12 Oddish (base 30 speed) Absorb:
98 hits 2 crits 0 misses
Expected rate: 5.86% Observed: 2%

Level 13 Slowpoke (base 15 speed) Confusion:
98 hits(2 with status) 2 crits 0 misses
Expected rate: 2.93% Observed: 2%

Level 13 Voltorb (base 100 speed) Tackle:
82 hits 18 crits 0 misses
Expected rate: 19.53% Observed: 18%

Looks like it's functioning how you want now, nice and lopsided :(

**Comment by jojobear13** [comment]
About every boss fight is fine. I'm probably going to go with a polynomial curve I came up with some time ago.

**Comment by DiabloStorm** [comment]
> About every boss fight is fine. I'm probably going to go with a polynomial curve I came up with some time ago.

Gym battles then and things like Giovanni? Alright.

**Comment by DiabloStorm** [comment]
Question: do trainer's that scale...scale DOWN? or are they set to their default pokemon level as default?

**Comment by jojobear13** [comment]
>or are they set to their default pokemon level as a minimum

That is correct. They do not scale down.

>Also, would you be open to making an optional battle scene (or different colors in battles/menus besides pure white) instead of the pure white snowblindness that currently exists?

I don't think this is technically feasible. The game boy only has a single background layer on which pokemon and the GUI elements are drawn.

**Comment by DiabloStorm** [comment]
"fixed-damage move fixes (seismic toss, dragon rage, etc):
can no longer critically hit
**obey type immunities**"

Of the random gastly I encountered a few days ago I think they were unable to use nightshade vs my spearow...but this move is supposed to ignore typing in the vanilla game.

**Comment by jojobear13** [comment]
>but this move is supposed to ignore typing in the vanilla game

No, it's not. Seismic toss and nightshade ignoring type immunity is a bug in the vanilla battle engine. 

>I don't understand your criteria, besides personal taste, for picking and choosing what to keep that was there by original intention

It involves doing a lot of forensics with the structure of the code.

>what to add from gen 2, what to not add from gen 2,

The only real features from gen 2 is identification of pokemon sex and shiny status. The only reason those are even in there is because those things already exist in gen 1 retroactively by virtue of being completely defined by a pokemon's DVs.

**Comment by DiabloStorm** [comment]
Do you have Discord? I feel if I keep asking questions I'm going to bury the data I'm adding to each time, making it harder to find. (Perhaps it should be moved or is there a way to merge comments here?)

I'm wondering if Hyper Beam is still a physical attack? Or was that an error and it's been changed to special?

**Comment by jojobear13** [comment]
I am on discord, but I do not have a public server. I typically use it for communicating with my collaborators (shinpokered, despite public appearances, is not entirely a one-man operation).

Github is good for archival, but you can always post things in shinpokered's pokecommunity thread.

>I'm wondering if Hyper Beam is still a physical attack?

This hack does not implement the gen 4 physical/special split. The normal-typed hyper beam is physical.

**Comment by DiabloStorm** [comment]
I moved the stuff you're probably more interested in seeing from me to the top to save all the scrolling :)

**Comment by DiabloStorm** [comment]
Are you sure EVs are set to 0 when the battle style has no !   ???

How is this gyarados tanking thundershock? it's 4x weak to it and pikachu (currently my highest level mon) only did like 1/5 of it's health

**Comment by DiabloStorm** [comment]
<details>
  <summary>What is this? 999?</summary>

![image](https://user-images.githubusercontent.com/63818362/210012650-57fd45ce-9f61-46cc-bb8f-dcfe6b1c93f3.png)

</details>

**Comment by DiabloStorm** [comment]
> About every boss fight is fine. I'm probably going to go with a polynomial curve I came up with some time ago.

Why is it you can't just scale EVs relative to the player's EVs instead of these approximations? If it's locked in some way to the way the player's team roster is progressing in EVs wouldn't this make it a bit of a fairer challenge?

**Comment by DiabloStorm** [comment]
> No, it's not. Seismic toss and nightshade ignoring type immunity is a bug in the vanilla battle engine.

I've been looking around on the usual sites...where are you getting this information from besides 'inferring from the code' ?

**Comment by DiabloStorm** [comment]
> 50% chance that the AI is blind to a player switching or using an item

I don't notice this 50% at all. It looks like it's 100%. Every time I switch in a type immune to the attack they keep using, they always use a different attack. you've got the AI straight up cheating, dude.

**Comment by DiabloStorm** [comment]
Out of curiosity I'm savestate testing the "RNG" using disable against a machamp, it's very obvious that something was messed with here. Disable refuses to work against Karate Chop because it is AI Machamp's preferred next move to use. You aren't going to explain this away by "bad luck." The only way I can get it to disable Karate Chop is with a bunch of my pokemon switching around until the AI favors another move, the moment it decides it wants karate chop again and it's disabled, ive noticed the game ends the disability basically immediately when I switch to another pokemon. 
I can't fully explain what's going on but I'm sure you can since you made it this way. Considering some of the files here are 4 years old I assume this project is at least that old, so there's been plenty of time to implement pervasive and not-so-immediately-obvious AI cheating like this.
[lol who does this? I'm starting to feel like this is some prank.](https://www.youtube.com/watch?v=pKlW-tqS6dE)

**Comment by jojobear13** [comment]
I'm going to be very frank here. Your hostile tone is not well-received by the associates of this project. Calling anyone a troll or a liar is inappropriate unless you fire up a debugger and post a code comparison that proves your case. This project is open source, as is the vanilla pokered project, with all the code available to be reviewed and compared between the two. There may (quite likely) be errors that have been induced, and some things might still be an exercise in learning, but there is nothing hidden in the code just to mess with people.

I see you are testing the outcomes of the RNG by constantly and rapidly resetting a save state. This is a flawed way to test outcomes. No digital RNG is truly random but is in actuality a complicated pattern that approximates randomness with a uniform distribution over time. When you reload the save state, you are also resetting the RNG back in time to an earlier point in the pattern. This causes the outcome from hitting the 'A' button (just after reloading) to fall within a finite gaussian distribution (instead of the expected uniform distribution) based on how long it took you to press the button. This is especially so if you keep reloading and hitting 'A' in some kind of rhythmic pattern that further constrains the place in time that an outcome is selected. Waiting some time to press the 'A' button will gradually flatten this gaussian into a uniform distribution because one new outcome is traversed in the pattern every 1/60th of a second. It can take at least 4 to 5 seconds for the 256 possible values to start looking like a uniform distribution again.

[Pokemon games are extremely vulnerable to inducing cognitive biases in players;](http://nuggetbridge.com/articles/common-cognitive-biases-competitive-pokemon/) the two big ones are the Gambler's Fallacy and Confirmation Bias. I'm seeing a lot of this in your posts. These are generally taken under consideration when someone posts about them or there is a pattern evolving from different playtesters. But verifying or debunking requires the time-sucking task of someone stepping through code in a debugger and mapping changes in register values. For example, there is [nothing fishy](https://github.com/jojobear13/shinpokered/blob/17e60a9d8e6c3ad8f3909e0df1640ae821d5ad41/custom_functions/func_enc_gen.asm#L281) with the randomization with regular trainers. It's a straight random pick from a list of unevolved pokemon species with the potential to evolve them if appropriate. It was even recently toned-down and rigged in-favor of the player prevent Scyther in Viridian forest or Gyarados on the SS Anne.

The AI improvements are a constant work in progress. Yes, the AI is omniscient. It has to be because the processor is from 1979, and that's how Gamefreak structured the battle engine. Perfect play is not an intended goal, but Gamefreak did not make it easy to keep this from happening. AI switching and ball-selecting is all-new, but code was added after playtesting to force the first enemy mon in the roster to be the first sendout (to prevent always sending out the best mon at the start of battle). AI Layer 1, which focuses mainly on moves other than straight damage, had to have a "blinding" feature added because enemies would, with 100% certainty, try to re-status the player the instant an antidote or the pokeflute or a switch-in was executed by the player. AI Layer 3, which handles type-matching, pays switching no mind because it runs with whatever the player has out on the field. And oh joy, Gamefreak did not program any kind of flag value that uniquely indicates the player switched. That's why in Yellow version you can guarantee that switching Blastoise into Lance's Dragonite will result in it precognitively choosing to Thunder. Commit 17e60a9d8e6c3ad8f3909e0df1640ae821d5ad41 is another attempt to do something about this (fingers crossed).

Adding stat exp to trainer pokemon has had its own challenges. Particularly CPU speed. Earlier attempts with matrix math or even just mass-calculate-and-copy the mean of the player DVs caused noticeable delay. 640 per level was added mainly as a "eh, it'll be fine" solution because it's very fast to do a simple linear multiply. A polynomial solution was commented out for a year or two because it wasn't working super great, but it's recently gotten a rewrite with some optimizations.

The attached zipped video shows that it is completely possible to preempt an enemy's favored move with the Disable effect.
[bgb-1672770579.zip](https://github.com/jojobear13/shinpokered/files/10339909/bgb-1672770579.zip)
I stepped through the code line-by-line in the debugger to make sure there were no errors in the structure of the function and that all logic states were accounted for.

As for the Disable turn counter, Disable has a very obscure bug in vanilla gen 1. It should disable the target for 1 to 8 turns, right? Wrong! If it hits a target that is going second in the round, and the counter for it gets set to 1 turn, then the disable effect will clear at the end of the round. Why? Because the Disable effect is initialized and its counter decremented **during** pokemon turn executions and **after** any kind of move selection. **But**, a slower opponent pokemon that gets hit with Disable already got to choose a move before getting hit with Disable, so the first turn of the effect gets wasted unless it was the chosen move itself that got Disabled. In most cases, Disable effectively lasts 0 to 7 turns on a slower target. 

Note that the recovery text box for Disable does not correspond to when the disable effect is doing its thing. If an enemy starts the round with a disable counter of 1, the recovery text box might make it look as if the effect immediately ended. This is not actually true because this text displays **after** the enemy has already gone through move-choosing procedures. A non-zero Disable counter will disallow the disabled move from being picked just the same as the player being unable to select it from the move menu.

Fixing such an obscure bug with Disable involved making functions that identify this peculiar state and that subsequently change the counter from a 1 into a 2 so as to prevent the effect from immediately ending. A faster pokemon getting hit with Disable has the following possible turn counter outcomes at the start of the next round: [1,2,3,4,5,6,7,8]. But now a slower pokemon getting hit with Disable starts the next round with one of the following outcomes [1,1,2,3,4,5,6,7]. This second distribution has a mean that skews -0.875 lower and is notably lopsided. **This is considered an insufficient solution, and improvements are slated to be made**.

**Comment by DiabloStorm** [comment]
Thanks for eventually getting back to me anyway.

My first run through this game was indeed naive, thinking I could play in a normal manner. The second run I feel I had a better understanding of what the game is demanding of me in its current state. Keep in mind in spite of my criticisms, I wouldn't be wasting my time complaining here if I felt this project was hopeless.


A partial list of some things I enjoyed:

Gender and caught indicator, great.
Rebattles, great. But implemented poorly at the difficulty I'm playing at.
Trainer scaling, great. But implemented poorly at the difficulty I'm playing at.
Option for random wild/trainer pokemon? Great, at face value.
Able to see DV/EVs? Godsend. (Though, I have an internal battle if the ensuing pokemon are actually "legal" if that matters...because you have enabled all 65535 DV spreads for all, right? bugfix I suppose)
Able to run, hotkeys for bike, rod, HM moves, HM slots for pokemon, all wonderful.
Catch up exp option? Great in theory. Unthinkable at the difficulty I'm playing at.
Backsprites look great.
Expanded backpack storage, great.
Instant text, great.
Payday/gamecorner interaction is nice.

Didn't get to explore chansey/shiny interaction yet, but sounds interesting
Same with the new missingno thing, seems interesting.
I'm I'd find more things I like as I go (and probably more annoyances too)

I can't dig through the code. 
You're getting the firsthand look at my experiences as a player between being completely naive to meta-gaming the system. I have no clue what all has been changed other than discovering as I play and reading here.
I was initially excited to play this after reading the entirety of the description here, I liked what I was reading. 
This hack still clearly needs a lot of work and I'm finding after 1 partial run, a restart, now past Celadon in a 2nd run, there is a list of things (mostly at the top here) I'm growing to dislike as crit moves and double team are effectively carrying me through the rest of the game.

Anyway, good luck. Hopefully something was gained other than "gee, this guy sure is critical". 

But, with that said, maybe this just isn't the romhack for me.
